canonical-websites/build.snapcraft.io
# Show extended error messages for snap builds with external link

## Done

Show extended error messages for snap builds with external link, if available (fixes [LP: #1729580](https://bugs.launchpad.net/build.snapcraft.io/+bug/1729580)).

Launchpad is now able to expose a list of errors for each failed snap build in the `store_upload_error_messages` field (the old `store_upload_error_message` field, which joined all error messages ignoring any links, is still available for backwards compatibility). Each object in `store_upload_error_messages` contains a `message` field and an optional `link`, as reported by the reviewer tools via the Snap Store. See e.g. [this snap build in Launchpad](https://launchpad.net/~maxiberta/+snap/40315ffd245d9e29b06194c82e7aae29-xenial/+build/154794).

## QA

- Check out this feature branch
- Run the site using the command ```npm start -- --env=environments/dev.env```
- View the site locally in your web browser at: [http://0.0.0.0:8000/](http://0.0.0.0:8000/)
- Verify any previous failed builds show the error message as usual.
- Build a snap that fails (e.g. by setting `confinement: classic` in `snapcraft.yaml`).
- Verify the new failed build shows each error separately, with a link if available.

## Issue / Card

Fixes [Use of classic confinement mentions "the forum" without further explanation (LP: #1729580)](https://bugs.launchpad.net/build.snapcraft.io/+bug/1729580).

## Screenshots

Before:
![screenshot-2018-2-23 build snapcraft io - maxiberta htop-snap-testing builds](https://user-images.githubusercontent.com/381336/36611588-c2c1171e-18b2-11e8-9424-2b965b2bfcfb.png)

After:
![screenshot-2018-2-23 build snapcraft io - maxiberta htop-snap-testing builds](https://user-images.githubusercontent.com/381336/36611503-6f7ad19e-18b2-11e8-85ae-56e7132fcf05.png)
